webui/react/src/components/DynamicTabs.tsx

```diff
@@ -0,0 +1,77 @@
+import { Tabs, TabsProps } from 'antd';
+import React, { createContext, useCallback, useContext, useEffect, useState } from 'react';
+import { useNavigate, useParams } from 'react-router-dom';
+
+import { isEqual } from 'shared/utils/data';
+
+interface DynamicTabBarProps extends Omit<TabsProps, 'activeKey'> {
+  basePath: string;
+}
+
+type TabBarUpdater = (node?: JSX.Element) => void;
+
+const TabBarContext = createContext<TabBarUpdater | undefined>(undefined);
+
+const DynamicTabs: React.FC<DynamicTabBarProps> = ({
+  basePath,
+  children,
+  ...props
+}): JSX.Element => {
+  const [tabBarExtraContent, setTabBarExtraContent] = useState<JSX.Element | undefined>();
+
+  const navigate = useNavigate();
+
+  const [tabKeys, setTabKeys] = useState<string[]>([]);
+
+  useEffect(() => {
+    const newTabKeys = React.Children.map(children, (c) => (c as { key: string })?.key ?? '');
+    if (Array.isArray(newTabKeys) && !isEqual(newTabKeys, tabKeys)) setTabKeys(newTabKeys);
+  }, [children, tabKeys]);
+
+  const { tab } = useParams<{ tab: string }>();
+
+  const [activeKey, setActiveKey] = useState(tab);
+
+  const handleTabSwitch = useCallback(
+    (key: string) => {
+      navigate(`${basePath}/${key}`);
+      setActiveKey(key);
+    },
+    [navigate, basePath],
+  );
+
+  useEffect(() => {
+    setActiveKey(tab);
+  }, [tab]);
+
+  useEffect(() => {
+    if (!activeKey && tabKeys.length) {
+      navigate(`${basePath}/${tabKeys[0]}`, { replace: true });
+    }
+  }, [activeKey, tabKeys, handleTabSwitch, basePath, navigate]);
+
+  const updateTabBarContent: TabBarUpdater = useCallback((content?: JSX.Element) => {
+    setTabBarExtraContent(content);
+  }, []);
+
+  return (
+    <TabBarContext.Provider value={updateTabBarContent}>
+      <Tabs
+        {...props}
+        activeKey={activeKey}
+        tabBarExtraContent={tabBarExtraContent}
+        onTabClick={handleTabSwitch}>
+        {children}
+      </Tabs>
+    </TabBarContext.Provider>
+  );
+};
+
+export default DynamicTabs;
+
+export const useSetDynamicTabBar = (content: JSX.Element): void => {
+  const updateTabBarContent = useContext(TabBarContext);
+  useEffect(() => {
+    updateTabBarContent?.(content);
+  }, [updateTabBarContent, content]);
+};
```

Comment:
allows us to get rid of duplicated logic around tab routing.

also provides an alternative way to set the contents of the tab bar from within a tab itself. the previous implementation was to have all the code for the tab content running in the parent component of the tabgroup

Reply:
Is this necessary for this feature or could it be moved to its own PR?

Reply:
this one in particular is needed to avoid having the experiment specific logic run in the parent component -- and therefore also when the experiment tab is inactive

Reply:
the difference between this and the other instances where we have different tables in different tabs is that this

<img width="229" alt="image" src="https://user-images.githubusercontent.com/28383080/195406705-f0d6e695-f7dd-4712-95d3-93659de51a16.png">

has to be specified as a prop on `Tabs` itself